### PR TITLE
Rosetta/sebp 82 support coin standard balance endpoint

### PR DESCRIPTION
**Pull request by nikos-kitmeridis**
## Description 

Adds support for coins other than SUI on the Rosetta`account/balance` endpoint
### Changes
* Adds support for multiple currencies on the same request
* SUI becomes the default currency on the request if none provided
* Refactor of the balance request method to reduce code duplication

## Test plan 
Unit and Integration tests

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: crates/sui-rosetta/src/types.rs [hunk 1]
```diff
@@ -95,18 +95,57 @@ impl From<SuiAddress> for AccountIdentifier {
     }
 }
 
+fn default_currency_metadata() -> Option<CurrencyMetadata> {
```
[the remark this answers is not in the document]
Comment (nikos-kitmeridis) [reply crates/sui-rosetta/src/types.rs]: @nikos-terzo I did replace (most of) these methods with Default trait implementations.

**Review by patrickkuo — COMMENTED** [review]
so now when currency is not specified, it is defaulted to SUI? 

Is this change ok with CB? I can't remember if rosetta require the currency info or not.

File: crates/sui-rosetta/src/account.rs [hunk 2]
```diff
@@ -29,108 +30,98 @@ pub async fn balance(
...
-                .client
-                .read_api()
-                .get_latest_checkpoint_sequence_number()
-                .await?;
+async fn get_checkpoint(ctx: &OnlineServerContext) -> CheckpointSequenceNumber {
+    ctx.client
+        .read_api()
+        .get_latest_checkpoint_sequence_number()
+        .await
+        .unwrap()
```
[the remark this answers is not in the document]
Comment (nikos-kitmeridis) [reply crates/sui-rosetta/src/account.rs]: @patrickkuo I've amended to return this one instead `SuiRpcResult<CheckpointSequenceNumber>`
is this what you were looking for more changes are required?

**Comment by nikos-kitmeridis** [comment]
> so now when currency is not specified, it is defaulted to SUI?
> 
> Is this change ok with CB? I can't remember if rosetta require the currency info or not.

So there has been a discussion and we concluded to the following addition of the `coin_type` field

> SUI would become:
"currency": {
  "symbol": "SUI",
  "decimals": 9,
  "metadata:
    {"coin_type":"0x2::sui::SUI"}
}
and WBTC would become
"currency": {
  "symbol": "WBTC",
  "decimals": 8,
  "metadata:
    {"coin_type":"0xb0b...a11ce::wbtc_module::WBTC"}
}

Which is ok for them but:
> except we do NOT want to require it for SUI. We want SUI to be inferred as it is today.

So we will default to SUI if the `metadata` field is not present on the request

File: crates/sui-rosetta/tests/custom_coins/test_coin_utils.rs [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+            digest,
+            owner,
+            ..
+        } = change
+        {
+            if object_type.to_string().contains("2::coin::TreasuryCap") {
+                // let TypeTag::Struct(coin_tag) = object_type.type_params.into_iter().next().unwrap()
+                // else {
+                //     return None;
+                // };
```
Comment (patrickkuo) [inline crates/sui-rosetta/tests/custom_coins/test_coin_utils.rs:195]: is this still needed?

File: crates/sui-rosetta/tests/custom_coins/test_coin/Move.toml [hunk 4]
```diff
@@ -0,0 +1,37 @@
+[package]
```
Comment (patrickkuo) [inline crates/sui-rosetta/tests/custom_coins/test_coin/Move.toml:1]: lets remove commented out template and unused sections

**Review by patrickkuo — APPROVED** [review]
some minor comments, otherwise LGTM

File: crates/sui-rosetta/src/account.rs [hunk 5]
```diff
@@ -29,108 +31,97 @@ pub async fn balance(
...
-                    balances: balances_first,
-                });
-            } else {
-                // retry logic needs to be aaded
-                retry_attempts -= 1;
-            }
+    while retry_attempts > 0 {
+        let balances_first = get_balances(&ctx, &request, address, currencies.clone()).await?;
+        let checkpoint1 = get_checkpoint(&ctx).await.unwrap();
+        let mut checkpoint2 = get_checkpoint(&ctx).await.unwrap();
```
Comment (manolisliolios) [inline crates/sui-rosetta/src/account.rs:39]: These unwraps are scary. I think they should all be `.await?` instead,
or a clear explanation why this unwrap cannot panic.

File: crates/sui-rosetta/src/account.rs [hunk 6]
```diff
@@ -29,108 +31,97 @@ pub async fn balance(
...
-                });
-            } else {
-                // balances are different so we need to try again.
-                info!(
-                    "different balance for account {} at checkpoint {}",
-                    address, checkpoint2
-                );
-                retry_attempts -= 1;
+        });
+        let balances: Vec<(Currency, Result<i128, Error>)> = join_all(balance_futures).await;
```
Comment (manolisliolios) [inline crates/sui-rosetta/src/account.rs:91]: maybe worth using try_join_all and error immediately?

**Pull request closed without merging**